File: ssh_key_manager.py
"""
SSH Key Manager Module
Handles SSH key generation, detection, and public key retrieval.
"""

import os
import subprocess
import platform
from pathlib import Path


class SSHKeyManager:
    """Manages SSH key operations including generation and retrieval."""
    
    def __init__(self):
        """Initialize SSH key manager with default paths."""
        self.home_dir = Path.home()
        self.ssh_dir = self.home_dir / ".ssh"
        self.private_key_path = self.ssh_dir / "id_rsa"
        self.public_key_path = self.ssh_dir / "id_rsa.pub"
    
    def check_ssh_keys_exist(self):
        """
        Check if SSH key pair already exists.
        
        Returns:
            bool: True if both private and public keys exist, False otherwise
        """
        return self.private_key_path.exists() and self.public_key_path.exists()
    
    def generate_ssh_key(self, email="", key_type="rsa", bits=2048):
        """
        Generate a new SSH key pair using ssh-keygen.
        
        Args:
            email (str): Optional email/comment for the key
            key_type (str): Type of key to generate (rsa, ed25519, etc.)
            bits (int): Number of bits for RSA keys
            
        Returns:
            tuple: (success: bool, message: str)
        """
        try:
            # Create .ssh directory if it doesn't exist
            self.ssh_dir.mkdir(parents=True, exist_ok=True)
            
            # Build ssh-keygen command
            cmd = [
                "ssh-keygen",
                "-t", key_type,
                "-b", str(bits),
                "-f", str(self.private_key_path),
                "-N", "",  # No passphrase
            ]
            
            if email:
                cmd.extend(["-C", email])
            
            # Execute ssh-keygen
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                creationflags=subprocess.CREATE_NO_WINDOW if platform.system() == "Windows" else 0
            )
            
            if result.returncode == 0:
                return True, "SSH key pair generated successfully!"
            else:
                error_msg = result.stderr.strip() or result.stdout.strip()
                return False, f"Error generating SSH key: {error_msg}"
                
        except FileNotFoundError:
            return False, "SSH client not found. Please install OpenSSH for Windows."
        except Exception as e:
            return False, f"Unexpected error: {str(e)}"
    
    def get_public_key(self):
        """
        Read and return the SSH public key content.
        
        Returns:
            tuple: (success: bool, key_content: str or error_message: str)
        """
        try:
            if not self.public_key_path.exists():
                return False, "Public key file not found. Please generate SSH keys first."
            
            with open(self.public_key_path, 'r') as f:
                public_key = f.read().strip()
            
            return True, public_key
            
        except Exception as e:
            return False, f"Error reading public key: {str(e)}"
    
    def get_private_key_path(self):
        """
        Get the absolute path to the private key file.
        
        Returns:
            str: Absolute path to private key
        """
        return str(self.private_key_path.absolute())
    
    def delete_existing_keys(self):
        """
        Delete existing SSH key pair (use with caution).
        
        Returns:
            tuple: (success: bool, message: str)
        """
        try:
            deleted = []
            
            if self.private_key_path.exists():
                self.private_key_path.unlink()
                deleted.append("private key")
            
            if self.public_key_path.exists():
                self.public_key_path.unlink()
                deleted.append("public key")
            
            if deleted:
                return True, f"Deleted: {', '.join(deleted)}"
            else:
                return True, "No keys to delete"
                
        except Exception as e:
            return False, f"Error deleting keys: {str(e)}"

File: Managers/SshKeyManager.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using System.Windows;

namespace IPTunnelGsmMeta.Managers
{
    /// <summary>
    /// Manages SSH key generation, detection, and operations
    /// </summary>
    public class SshKeyManager
    {
        private readonly string _sshDir;
        private readonly string _privateKeyPath;
        private readonly string _publicKeyPath;

        public SshKeyManager()
        {
            _sshDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".ssh");
            _privateKeyPath = Path.Combine(_sshDir, "id_rsa");
            _publicKeyPath = Path.Combine(_sshDir, "id_rsa.pub");
        }

        /// <summary>
        /// Check if SSH keys exist
        /// </summary>
        public bool KeysExist()
        {
            return File.Exists(_privateKeyPath) && File.Exists(_publicKeyPath);
        }

        /// <summary>
        /// Generate new SSH key pair
        /// </summary>
        public async Task<(bool success, string message)> GenerateKeyAsync(string comment = "")
        {
            try
            {
                // Ensure .ssh directory exists
                Directory.CreateDirectory(_sshDir);

                if (KeysExist())
                {
                    return (false, "SSH keys already exist. Delete existing keys first if you want to regenerate.");
                }

                // Build ssh-keygen command
                var emailComment = string.IsNullOrEmpty(comment) ? "tech@gsmmeta.com" : comment;
                var arguments = $"-t rsa -b 2048 -f \"{_privateKeyPath}\" -N \"\" -C \"{emailComment}\"";

                var processStartInfo = new ProcessStartInfo
                {
                    FileName = "ssh-keygen",
                    Arguments = arguments,
                    UseShellExecute = false,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    CreateNoWindow = true
                };

                using var process = Process.Start(processStartInfo);
                if (process == null)
                {
                    return (false, "Failed to start ssh-keygen process");
                }

                var output = await process.StandardOutput.ReadToEndAsync();
                var error = await process.StandardError.ReadToEndAsync();
                await process.WaitForExitAsync();

                if (process.ExitCode == 0 && KeysExist())
                {
                    return (true, "SSH keys generated successfully!");
                }
                else
                {
                    return (false, $"Key generation failed: {error}");
                }
            }
            catch (Exception ex)
            {
                return (false, $"Error generating SSH keys: {ex.Message}");
            }
        }

        /// <summary>
        /// Get the public key content
        /// </summary>
        public string? GetPublicKey()
        {
            try
            {
                if (File.Exists(_publicKeyPath))
                {
                    return File.ReadAllText(_publicKeyPath).Trim();
                }
                return null;
            }
            catch
            {
                return null;
            }
        }

        /// <summary>
        /// Copy public key to clipboard
        /// </summary>
        public bool CopyPublicKeyToClipboard()
        {
            try
            {
                var publicKey = GetPublicKey();
                if (publicKey != null)
                {
                    Clipboard.SetText(publicKey);
                    return true;
                }
                return false;
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        /// Get key information
        /// </summary>
        public string GetKeyInfo()
        {
            if (!KeysExist())
            {
                return "No SSH keys found";
            }

            try
            {
                var fileInfo = new FileInfo(_privateKeyPath);
                return $"Key created: {fileInfo.CreationTime:yyyy-MM-dd HH:mm}";
            }
            catch
            {
                return "SSH keys exist";
            }
        }

        /// <summary>
        /// Delete existing SSH keys
        /// </summary>
        public (bool success, string message) DeleteKeys()
        {
            try
            {
                if (File.Exists(_privateKeyPath))
                {
                    File.Delete(_privateKeyPath);
                }
                if (File.Exists(_publicKeyPath))
                {
                    File.Delete(_publicKeyPath);
                }
                return (true, "SSH keys deleted successfully");
            }
            catch (Exception ex)
            {
                return (false, $"Error deleting keys: {ex.Message}");
            }
        }
    }
}
